### PR TITLE
loader: explicitly manage policy programs in the Cilium agent

### LINKED ISSUE #29331: Endpoint creation should explicitly insert policy programs into cilium_(egress)calls_policy
As fixed and documented in https://github.com/cilium/cilium/pull/29307, relying on the ELF loader to insert programs into `cilium_(egress)calls_policy` is a bad idea, since the order of populating BPF maps is random. Inserting a prog into a policy call map should be considered equivalent to an attachment to a BPF hook, and should only be done after an ELF's maps have been fully populated.

This issue is for:
1. lifting prog insertion out of the loader package into code responsible for creating lxc and host endpoints.
1. convert `POLICY_*_MAP` to BTF map definitions since they no longer need populating by the loader
1. remove `__section_tail` annotations from `handle_policy()`, `handle_policy_egress()` and `handle_lxc_traffic()`
1. removing ELF substitutions for the above progs' section names

The above points likely mean moving away from `replaceDatapath()` since access to the Collection (or a `LoadAndAssign()` object) is needed for getting handles to the policy programs.

### PR DESCRIPTION
**Pull request by ti-mo**
Fixes: #29331

This PR puts the agent in charge of picking out specific programs from a Collection and inserting them into a policy map at a defined, predictable point during the loading process. All but one remaining (`cilium_calls_`) bpf_elf_map has been converted to BTF-style map definitions. The last step in sunsetting the iproute2 legacy is porting over cilium_calls, which is now well within reach.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -72,6 +72,9 @@ test/cilium-[0-9a-f]*.yaml
 test/*tmp
 test/cilium-istioctl
 
+# Testdata, includes generated .o
+!**/testdata/*
```
Comment (bimmlerd) [inline .gitignore:76]: Is there no way we can avoid committing these binary files?

Comment (ti-mo) [reply .gitignore:76]: Good question! Stuffing binary files in Git is generally frowned upon, but:
- These files are generated by a single command using a stable, Dockerized toolchain
- CI always clean-builds the files to ensure there's no diff
- They're stripped as far down as possible, leaving just the ELF structure, BTF blob and the binary contents of the prog and map sections. The objects in this repo are max 4000 bytes large. For reference, this repo's vendor/ is currently a whopping 162 MiB of source code.

This buys us lightning-fast `go test` runs without any runtime toolchain or other external dependencies, a trade-off that's well worth it.

File: bpf/lib/maps.h [hunk 2]
```diff
@@ -134,14 +134,13 @@ struct bpf_elf_map __section_maps CALLS_MAP = {
...
  */
-struct bpf_elf_map __section_maps CUSTOM_CALLS_MAP = {
-	.type		= BPF_MAP_TYPE_PROG_ARRAY,
-	.id		= CILIUM_MAP_CUSTOM_CALLS,
-	.size_key	= sizeof(__u32),
-	.size_value	= sizeof(__u32),
-	.pinning	= LIBBPF_PIN_BY_NAME,
-	.max_elem	= 4,	/* ingress and egress, IPv4 and IPv6 */
-};
+struct {
```
Comment (rgo3) [inline bpf/lib/maps.h:134]: Reading the respective commit message of the change, what's the risk for users when changing maps to a BTF definition?

**Review by bimmlerd — APPROVED** [review]

File: pkg/datapath/loader/loader.go [hunk 3]
```diff
@@ -481,6 +469,23 @@ func (l *loader) reloadDatapath(ep datapath.Endpoint, spec *ebpf.CollectionSpec)
 		}
 		defer coll.Close()
 
+		// Insert policy programs before attaching entrypoints to tc hooks.
```
Comment (rgo3) [inline pkg/datapath/loader/loader.go]: General question: During upgrade, does this update maps that are used by previously attached programs? And what are the risks of previously compiled tc attachments calling into newly compiled tail calls?

**Review by rgo3 — APPROVED** [review]
Overall LGTM, just two general questions about the code.  `LoadAndAssign` improves the readability by a lot :+1:

File: bpf/lib/maps.h (hunk 2 above)
Comment (ti-mo) [reply bpf/lib/maps.h:134]: None! It's just a different representation in C, with the added benefit that the new definition will cause MapSpec.Key and .Value to be populated with btf.Type values corresponding to u32s, leading to `bpftool dump map` displaying pretty-prints. :)

File: pkg/datapath/loader/loader.go (hunk 3 above)
Comment (ti-mo) [reply pkg/datapath/loader/loader.go]: It does, `obj.PolicyMap` points to the system-global `cilium_call_policy`, which the `.Update()` call puts an entry into. This doesn't only happen on upgrade, but on every endpoint regeneration (due to policy or config changes at runtime).

I think the code comment already provides a high-level description, but the risks of old Collections calling into new Collections heavily depend on the bpf C code in question, and I can't speak to that specifically. Programs in the policy map are called when e.g. the program evaluating cilium_host egress traffic towards a pod needs to do a policy check for that pod. This effectively crosses the boundary from bpf_host into a policy program living in bpf_lxc. When this happens, we need to make sure the target bpf_lxc's internal state (mostly its cilium_calls) is fully populated with the new programs so we don't hit any missed tail calls. As long as that's the case, we're good. :)

**Review by markpash — APPROVED** [review]

**Review by glibsm — APPROVED** [review]

**Pull request merged**